kbase-extension/static/kbase/js/widgets/appWidgets2/input/selectInput.js

```diff
@@ -19,16 +19,34 @@ define([
         select = t('select'),
         option = t('option');
 
+    /**
+     *
+     * @param {object} config has fields:
+     *  - parameterSpec - object - the spec object with parameter info as built by the Spec object
+     *  - availableValues - optional Array of objects - if given, this supercedes the list of values given in the parameterSpec.
+     *      as in the parameterSpec.data.constraints.options, each should be an object with structure:
+     *      {
+     *        display: string - the string to display as an option
+     *        value: string - the value to set when selected
+     *      }
+     *  - initialValue - string - the value that should be selected when started
+     *  - disabledValues - array - the values that should be disabled at start (if initialValue is here, it's ignored)
+     *  - channelName - string - the bus channel to use
+     *  - showOwnMessages - boolean - if true, this widget shows its own messages (better description to come)
+     * @returns
+     */
     function factory(config) {
         const spec = config.parameterSpec,
             runtime = Runtime.make(),
             busConnection = runtime.bus().connect(),
             channel = busConnection.channel(config.channelName),
             model = {
-                availableValues: spec.data.constraints.options,
+                availableValues: config.availableValues || spec.data.constraints.options,
                 value: config.initialValue,
+                disabledValues: new Set(config.disabledValues || []),
```

Comment:
Making these `Sets` so they're faster lookup times (assuming lots of file inputs), and pretty broadly browser-friendly at this point (I don't think there's a Set polyfill set up) - https://caniuse.com/?search=new%20Set

Reply:
If it's a problem, the [](MDN Set Page) recommends [https://github.com/zloirock/core-js#set](this polyfill)

Reply:
I'm fairly sure we have sets elsewhere anyway.